### ISSUE TITLE
[rust-server] how to cope with loss of `file` type?

### ISSUE THREAD

**Issue by bjgill** [OpenAPITools/openapi-generator#538]
##### Description

Swagger v2 had three types to represent binary data: `{type: string, format: binary}`, `{type: string, format: byte}`, and `{type: file}`. In OpenAPI v3, we lost the last of these.

In `rust-server`, binary and byte were treated identically - as `swagger::ByteArray`. File was `Box<Future<Item=Option<Box<Stream<Item=Vec<u8>, Error=Error> + Send>>, Error=Error> + Send>`.

I'm not sure what the correct approach for OpenAPI v3 is. Should we keep one of binary/byte as `swagger::ByteArray` and move the other to use the same type as file used to? Do we keep binary/byte as-is and add in some `rust-server`-specific workaround for those wanting to retain the old file behaviour?

##### Related issues/PRs

Discovered in https://github.com/OpenAPITools/openapi-generator/issues/307

**Comment by bjgill** [comment]
Right - I've spoken with some of the people using swagger-codegen at the moment, and it turns out that there are currently people using the `file` type. It's necessary when transferring large files to avoid excessive memory usage.

Broadly, I think there are 3 options:
1. Drop support for `swagger::ByteArray`, and use the stream option for `binary`/`byte`. This has the advantage of simplicity, but is really clunky to use for those who just want a `Vec<u8>`
2. Keep the stream-based mechanism for large files, but hide it behind a format or custom property. Simple to implement. However, given that this choice is completely transparent to peers, the openAPI spec seems like the wrong place to be making this choice.
3. Define a new type/trait for `binary`/`byte` APIs that can be converted to/from both streams and `Vec<u8>`. This gives control over to the person integrating with the autogenerated code. Having a new type/trait will also give us a natural place to document this behaviour.

I personally prefer the 3rd option. However, I'll be talking to a number of other users to get some feedback. I'm expecting to have a design ready for implementation by the end of the week.

**Comment by bjgill** [comment]
I've now had some discussions. This revealed a few things.

1. First, the old `Stream`-based file handling wasn't actually streaming on the server - we buffered the file in its entirety, and then presented the in-memory file as a stream. 
2. Second, there is a rather gnarly design problem server-side. Things are simple for the client - it can just push files out over the wire in whatever format it prefers. However, if there is more than 1 parameter, the server needs to know what order to expect the parameters in to avoid needing to buffer the whole request until it can work out the order for itself. 
3. The worst case would be a request with multiple files. Imagine a server was expecting files A and B, and got file A first. Then, it forwarded the request to a client that wanted to send B first. This would lead to a deadlock.

It seems as if there is going to need to be quite a bit of design work before we can handle files properly without needing to buffer them to memory.

Therefore, given that we currently buffer files to memory anyway, I think the best solution is to remove the `Stream`-based API now. This will give us a `rust-server` that is fully feature-compatible with the version in swagger-codegen. I'll then raise a new issue to handle the problem of handling files asynchronously. When we find someone who actually needs this feature, we/they can then start thinking about solving the design problems.

I'm therefore going to revive #547. Please shout within the next day if this is going to cause problems for you.

**Comment by bossmc** [comment]
On (2), I'm assuming you're mostly talking about multipart bodies when you talk about ordering?  In which case I think supporting out-of-order bodies is pretty much par for the course for the server-side implementation and exposing them to the server in the order they arrive is fine, so long as there's a way for the server to re-order them if needed (by caching the contents in local memory or even on disk until the needed file arrives).

**Comment by bjgill** [comment]
That sounds correct, yes. We'd probably want to give some thought to the API design, though, to make it harder for server implementers to shoot themselves in the foot.

**Comment by bjgill** [comment]
We've now implemented the removal in #547. I'm going to close this PR on that basis. When someone needs file support, they can open a new issue.

**Comment by RobbieMcKinstry** [comment]
@bjgill By chance, do you know if streaming file support was ever added back? I don't see an open issue, and this is the only closed issue that I could find.

Thanks for your work on this project. ❤️

**Comment by RobbieMcKinstry** [comment]
Thank you very much :)
